Dynatrace/dynatrace-operator
# Optionally parallel downloading csi-provisioner

# Description

Introduces a way to run multiple reconciles in parallel in the csi-provisioner, hidden behind an ENVVAR (`MAX_PARALLEL_RECONCILES`).

This is trying to speed up the startup incase multiple dynakubes with different tenants are deployed at the same time all needing the csi driver to mount volumes.

### Edge cases we check for:
- Out of range values for `MAX_PARALLEL_RECONCILES` (should be between 1 - 40)
- We requeue a dynakube reconcile request if:
  - we reached the max defined in the envvar for parallel reconciles running at the same time
  - we are already reconciling the requested dynakube in one of the parallel reconciles

### Gotchas:
- The default memoryLimit and cpuLimit were meant for 1 download at a time and not `x`, so those probably need to be raised at high number of parallel downloads
  - memoryLimit can cause `CrashLoopBackOff` because of `OOMKilled`
  - cpuLimit can cause the downloads to be slow, especially at the unzip part. (the pods wont get killed but kubernetes wont give the necessary cpu to do the job in time)
- Error handling suffers a bit, due to the fact that the framework we use for reconciliation is not meant to be used this way, so it can happen that something is failing in the provisioner yet we can only find it out from the logs of the container. 

## How can this be tested?
1. Have a cluster (that didn't have csi-drivers on them before to make sure that we always have to download the binaries)
2. Modify the csi daemonset yaml to have the `MAX_PARALLEL_RECONCILES` set to something more then "1"
3. Deploy as many dynakubes with different tenants as you can do. (`applicationMonitoring` requires less setup, don't forget to use the `useCSIDriver: true`)
4. Deploy sample apps at the same time for all the dynakubes. (be aware of the `namespaceSelector`)

Expected result: 
Parallel downloads for each tenant (grep for `parallel` in the logs)

## Checklist
